### PR TITLE
Revert AbstractTask deletion and restore binary compatibility with groovy plugins

### PR DESCRIPTION
**Pull request by donat**
In Gradle 7.0, we tried to remove `AbstractTask`. To provide backward compatibility we introduced runtime bytecode manipulation to reroute any `AbstractTask` access to `DefaultTask`.

Unfortunately, there are cases where this approach fails. We've received a [stack overflow report ](https://github.com/gradle/gradle/issues/16199)that was due to this change. After assessing the possibilities we decided to roll back with the removal.

### CODE REVIEW

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy [hunk 1]
```diff
@@ -93,6 +93,8 @@ abstract class AbstractNativeParallelIntegrationTest extends AbstractInstalledTo
         setupParallelTaskAndExpectations(taskName)
 
         buildFile << """
+            import org.gradle.api.internal.AbstractTask
```
Comment (eskatos) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy:96]: This is unused

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java [hunk 2]
```diff
@@ -109,6 +111,15 @@
 
     private static final String[] NO_EXCEPTIONS = new String[0];
 
+    private static final List<Pair<String, String>> RENAMED_INTERFACES = asList(
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java:114]: :x: we should not revert the changes to this type except the presence of `AbstractTask` in this list. We should keep the extracted `InstrumentingBackwardsCompatibilityVisitor.java`.

**Comment by donat** [comment]
@bot-gradle test qF

**Comment by donat** [comment]
@bot-gradle test qFLO

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java (hunk 2 above)
Comment (donat) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java]: Fixed.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy (hunk 1 above)
Comment (donat) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy]: Removed

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java [hunk 3]
```diff
@@ -128,7 +128,7 @@ public void applyConfigurationTo(Hasher hasher) {
         private boolean isInterface;
 
         public InstrumentingVisitor(ClassVisitor visitor) {
-            super(ASM_LEVEL, visitor);
+            super(ASM7, visitor);
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java:131]: :x: let's keep using `ASM_LEVEL` here

Comment (donat) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java]: Fixed

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@eskatos would it make sense to add coverage for the original SOF error? (see https://github.com/gradle/gradle/commit/a84e462827c272e1c42b83d5ac63da61ead718cb)

**Comment by eskatos** [comment]
@donat some coverage sounds good yet

**Comment by donat** [comment]
@eskatos The test is part of this PR: https://github.com/gradle/gradle/pull/16262/files#diff-6072fcc98cc379ae285fc11b7b228a0cb59ffe2b38fb79c18ee3bd14887ab51aR216

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Review by eskatos — APPROVED** [review]

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle test RFN

**Pull request merged**